sdk/typescript/runtime/main.go

```diff
@@ -340,17 +375,138 @@ func (t *TypescriptSdk) detectRuntime(ctx context.Context, modSource *ModuleSour
+		if slices.Contains(entries, "bun.lockb") {
+			return Bun, "", nil
+		}
+
 		if slices.Contains(entries, "package-lock.json") ||
 			slices.Contains(entries, "yarn.lock") ||
 			slices.Contains(entries, "pnpm-lock.yaml") {
 			return Node, "", nil
 		}
-		if slices.Contains(entries, "bun.lockb") {
-			return Bun, "", nil
-		}
 	}
 
 	// Default to node
 	return Node, "", nil
 }
+
+// detectPackageManager detects the package manager from the user's module.
+// If the package.json file has a field "packageManager", it will use that to
+// determine the package manager to use. Otherwise, it will use the default
+// package manager based on the lock files present in the module.
+//
+// If none of the above works, it will use yarn.
+//
+// Except if the package.json has an invalid value in field "packageManager", this
+// function should never return an error.
+func (t *TypescriptSdk) detectPackageManager(ctx context.Context, modSource *ModuleSource, subPath string, runtime SupportedTSRuntime) (SupportedPackageManager, string, error) {
+	// If the runtime is Bun, we should use BunManager
+	if runtime == Bun {
+		return BunManager, "", nil
+	}
+
+	// Try to detect package manager from package.json
+	source := modSource.ContextDirectory().Directory(subPath)
+
+	// read contents of package.json
+	json, err := source.File("package.json").Contents(ctx)
+	if err == nil {
+		// Check if the dagger.runtime field is set to "bun" first
+		value := gjson.Get(json, "dagger.runtime").String()
+		if value != "" && strings.Contains(value, "bun") {
+			return BunManager, "", nil
+		}
+
+		value = gjson.Get(json, "packageManager").String()
+		if value != "" {
+			// Retrieve the package manager and version from the value (e.g., yarn@4.2.0, pnpm@8.5.1)
+			packageManager, version, _ := strings.Cut(value, "@")
+
+			switch SupportedPackageManager(packageManager) {
+			case Yarn, Pnpm, Npm:
+				return SupportedPackageManager(packageManager), version, nil
+			default:
+				return "", "", fmt.Errorf("detected unknown package manager: %s", packageManager)
+			}
+		}
+	}
+
+	// Try to detect package manager from lock files
+	entries, err := source.Entries(ctx)
+	if err == nil {
+		if slices.Contains(entries, "bun.lockb") {
+			return BunManager, "", nil
+		}
+
+		if slices.Contains(entries, "package-lock.json") {
+			return Npm, "", nil
+		}
+
+		if slices.Contains(entries, "yarn.lock") {
+			return Yarn, "", nil
+		}
+
+		if slices.Contains(entries, "pnpm-lock.yaml") {
+			return Pnpm, "", nil
+		}
+	}
+
+	return Yarn, YarnDefaultVersion, nil
+}
+
+// generateLockFile generate a lock file for the matching package manager.
+func generateLockFile(ctx context.Context, ctr *Container, packageManager SupportedPackageManager, version string) (*Container, error) {
+	switch packageManager {
+	case Yarn:
+		// Sadly, yarn < v3 doesn't support generating a lockfile without installing the dependencies.
+		// So we use npm to generate the lockfile and then import it into yarn.
```

Comment:
What if yarn `version` is >1?

Reply:
then it's possible to generate the lockfile, but it would have the same result that this method so I prefer keeping it that way,it cover more cases

Reply:
You're assuming the result is the same, but ok.